### PR TITLE
refactor(#229): API data validation and filtering logic

### LINKED ISSUE #254: Refactor Filters
## Summary

Implement filters according to ADR decisions.

## Acceptance Criteria

- [ ] Filters refactored
- [ ] Tests pass
- [ ] Add filters tests for boundary cases

## Notes

FYI: @stasiaaleks

Filter tests are needed for this week lab

### PR DESCRIPTION
**Pull request by stasiaaleks**
-> from new ADR 0005
 
 ## Context

During SRE review, we discovered several issues:

- Current API incoming data validation is not consistent. Filters are validated in API, service and repository layers. This breaks separation of concerns and makes it unclear where to catch proper validation errors. This applies to checking immutable fields, guardrail-ing pagination parameters etc.
- Lower-level repositories are not catching database errors. On the contrary, they are ignored up to API layer and are caught here. It couples database errors handling logic to the API layer.
- Sometimes service and repository are ignored and database methods are called directly in the API layer.

## Decision

1. **Introduce Pydantic models for API input validation**. Client-provided data is validated at the highest level of abstraction - the API layer - and then passed as a correct, validated object to lower layers.
   - Separate Pydantic models for different operations (create, update, patch) to enforce immutability at schema level
   - Use `model_validate()` for automatic type coercion and validation
   - Validation errors are caught and re-raised as DRF ValidationError with detailed error context

2. **Catch database errors in the repository layer** and propagate model-specific error wrappers to higher layers.
   - Repository methods wrap Django ORM exceptions (e.g., `IntegrityError`, `DoesNotExist`) into domain-specific exceptions
   - Domain exceptions inherit from DRF exception classes for automatic HTTP status code mapping
   - This prevents database implementation details from leaking into service and API layers

### Positive

- Less error-prone code due to automatic type coercion and validation
- Clear separation of concerns: validation at API boundary, business logic in services, data access in repositories
- Type-safe data flow throughout the application stack
- Reduced code duplication through reusable validation models
- Better error messages with structured validation feedback
- Immutable fields protected at schema level (no manual filtering needed)

### Negative

- Additional dependency (Pydantic) alongside DRF serializers
- DRF serializers still used for response serialization and API schema generation

### Additional Improvements Made During Refactoring

- **Decorator pattern for authorization**: `@require_student` and `@require_rating_ownership` decorators eliminate duplicate ownership checking code (~60% code reduction)
- **Custom exception hierarchy**: Domain-specific exceptions (e.g., `RatingNotFoundError`, `OnlyStudentActionAllowedError`) provide clear error semantics and automatic HTTP status mapping
- **Service layer refactoring**: Entity-specific services implement `IFilterable` protocol for reusable filter options, adhering to Single Responsibility Principle
- **Type safety improvements**: Protocol interfaces (`IRating`, `ICourse`) decouple business logic from Django ORM models; explicit type hints for decorator-injected parameters
- **Structured logging**: Consistent logging with `structlog` for better observability and audit trails
 
 
 ## Notes
 I have not yet found how to incorporate Pydantic schemas into API schema generation. Even though I have tried the way used here https://drf-spectacular.readthedocs.io/en/latest/blueprints.html#pydantic , it did not work and I am still working on a solution. I will create another PR for it
 
 ## Testing - I have contacted @katerynabratiuk and sent the list of tests that can be used in the report
 
Resolves #254 

## Summary by CodeRabbit

* **New Features**
  * Richer course & rating filtering (sort by averages, semester/year options) and consistent paginated search responses.

* **Bug Fixes**
  * Standardized validation, clearer not-found/identifier/conflict responses, and stricter enrollment/ownership checks.

* **Documentation**
  * Added API data-validation ADR and improved OpenAPI parameter descriptions.

* **Refactor**
  * Services, repositories and pagination rewired to schema-driven validation and dependency injection.

* **Tests**
  * Large suite of new unit/integration tests covering filtering, deduplication, grouping and rating flows.

* **Chores**
  * Env tweaks and Python version bump.

### CODE REVIEW

File: src/backend/rating_app/repositories/rating_repository.py [hunk 1]
```diff
@@ -56,30 +52,44 @@ def filter(
...
+        update_data_map = update_data.model_dump(exclude_none=True)
+        for attr, value in update_data_map.items():
+            setattr(rating, attr, value)
 
-    def create(self, **rating_data) -> Rating:
-        return Rating.objects.create(**rating_data)
+        try:
+            rating.save()
+        except IntegrityError as err:
+            raise DuplicateRatingException() from err
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/rating_app/repositories/rating_repository.py]: nice moment, will resolve it

File: src/backend/scraper/services/deduplication/extractors.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+class RequiredFieldExtractor(Extractor[ParsedCourseDetails, str]):
+    def __init__(self, field_name: str):
+        self.field_name = field_name
+
+    def extract(self, data: ParsedCourseDetails) -> str:
+        value = getattr(data, self.field_name, None)
+        if value is None or (isinstance(value, str) and not value.strip()):
+            raise DataValidationError(f"Course {data.id} missing required {self.field_name}")
+        return value.strip() if isinstance(value, str) else value
```
[the remark this answers is not in the document]
Comment (stasiaaleks) [reply src/backend/scraper/services/deduplication/extractors.py]: rebase error, not applicable to the task

**Review by Fybex — APPROVED** [review]
